### PR TITLE
Override prior type detection for embedded puppet (closes: #127)

### PR DESCRIPTION
**Pull request by lelutin**
Currently, if an `.epp` file starts with a code opening tag, `<%`, it gets
auto-detected as being of the `mason` type by Vim's core type detection,
which runs before ftdetect scripts are interpreted. The `.epp` extension
is not used by anything else, so we want to make sure that the core
autodetection doesn't take effect here.

To also document the other exception to file type detection in our
plugin, the use of `au!` is necessary for `.pp` file since the core type
detection already assigns them as pascal. Let's make a note as to why
we're using this forceful command in this particular case.

The effect of the `!` after the `au` keyword is that all autocommands get
removed and then this one gets added. This could potentially hide some
other autocommands that would be supplied by other plugins or user
scripts that get loaded before this plugin.

There should not be a need for an override of any type for Puppetfile,
so let's stop using the `!` suffix for that case.

### CODE REVIEW

**Comment by lelutin** [comment]
this iteration works better than my previous one. it solves the issue with .epp files and clarifies why we have those overrides in filetype detection

**Comment by lelutin** [comment]
I've rebased this on top of master now that the nvim tests are fixed.

**Comment by lelutin** [comment]
I've looked over in vader.vim and trawled through a couple of the projects in vader's wiki page about which projects are using it and I don't think that we can test filetype detection using vader unfortunately.  the content in the `Given` blocks are not passed through file type detection and we're testing without a file with an predictable name.

**Comment by lelutin** [comment]
@rodjek do you have an idea of how we could test filetype detection for puppet and embeddedpuppet files?

**Comment by lelutin** [comment]
ok now we have a test case for this. I'm not super happy with the test since it creates and then deletes a file in the git repository but I don't think vader has the means to help us to do better than this :\

I'll wait for a tiny bit of time to get feedback on the PR.

File: test/filetype/embeddedpuppet.vader [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Given (embeddedpuppet file with leading tag):
+  <% if { %>
+  content
+
+# This is super unclean: it creates a file at the top of the git repository. We
+# have to hope that a file with this name doesn't already exist. However, vader
+# doesn't seem to have the concept of a temporary directory for test runtime.
+Execute (create file and clear filetype):
+  write! test_with_leading_tag.epp
+  set ft=txt
```
Comment (dkearns) [inline test/filetype/embeddedpuppet.vader:14]: You could avoid writing a file here by using :file to name the buffer and then rerunning filetype detection.
```
  file test_with_leading_tag.epp
  filetype detect
```

File: test/filetype/embeddedpuppet.vader [hunk 2]
```diff
@@ -0,0 +1,22 @@
+Execute (Filetype detection on a new empty file):
+  edit foo.epp
+  AssertEqual &filetype, 'embeddedpuppet'
```
Comment (dkearns) [inline test/filetype/embeddedpuppet.vader:3]: These assertion args should be reversed - `<expected>, <got>`

File: test/filetype/embeddedpuppet.vader [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  write! test_with_leading_tag.epp
+  set ft=txt
+
+# Here we delete the file after opening it. We should have the filetype
+# detected so we don't need the file anymore. If we wait after the assertion, the
+# file won't get deleted if we failed the test.
+Then (Filetype detection with an opening epp tag at the start of the file):
+  edit test_with_leading_tag.epp
+  call delete('test_with_leading_tag.epp')
+  AssertEqual 'embeddedpuppet', &filetype
```
Comment (dkearns) [inline test/filetype/embeddedpuppet.vader:22]: Leaving just the assertion here in the `Then` block.

File: test/filetype/embeddedpuppet.vader (hunk 2 above)
Comment (lelutin) [reply test/filetype/embeddedpuppet.vader:3]: you're right, that's how vader's readme documents it. however its output is misleading. this is what it says when there's an error and it's in the documented order:

~~~
    (2/2) [   THEN] (X) 'embeddedpuppet' should be equal to 'mason'
~~~

but you're right, I'll change it for the documented order

File: test/filetype/embeddedpuppet.vader (hunk 1 above)
Comment (lelutin) [reply test/filetype/embeddedpuppet.vader:14]: wow thanks that's way better than what I wrote and it seems to work for our testing purposes :)

**Comment by lelutin** [comment]
I've integrated the fix to the test you suggested and rebased on top of master so that I could reverse the order of commits: first add a test, then fix the problem.

I think with this bit of change we should probably be ready to merge this in, provided that CI is happy with the current state.

**Comment by lelutin** [comment]
with that trick you showed me, I couldn't help but add simple filetype detection tests for .pp and Puppetfile files too, for good measure.

**Comment by lelutin** [comment]
ah interesting .. I thought I was going mad since the tests were failing for every commit on this branch when I was running them locally -- even though the filetype detection is working correctly -- so I pushed to see if travis would obtain the same results as I was.

I currently don't really understand why the tests are failing though.. When editing files normally, detection does work correctly. Also when I implemented the tests as suggested with `filetype detect`, I remember that they were working locally.

**Comment by lelutin** [comment]
ah .. one interesting bit of clue came up when I tried something: if I edit the Puppetfile of a control repository, it gets detected as ruby.

but then if I run the following, the filetype reverts back to `conf` !

~~~
:filteype detect
~~~

so now that I'm able to mimic the error in the tests, I'll have to figure out why it's doing something different than when vim loads up or opens up a Puppetfile for edition after startup (in both cases the file gets detected correctly) and unbreak the tests

**Comment by lelutin** [comment]
I've amended the last commit since vim's documentation says that files in ftdetect don't need to declare an augroup. we should rather disable the vint check for that particular condition in this file.

it does not fix the tests, but hopefully I'll end up finding out how to get there soon

**Comment by lelutin** [comment]
apparently for Puppetfile, if there is a comment line at the start of the file, `filetype detect` intercepts detection and the filetype becomes `conf` ... again with vim's default type detection that comes in before plugins gah why ..

but also.. why is this not happening consitently when editing a file with a comment line at the start?

also if I remove the comment line at the start once I'm stuck in `conf` filetype and if I run again `:filetype detect`, vim doesn't go back to the `ruby` filetype. this is getting annoyingly inconsistent

**Comment by lelutin** [comment]
exactly the same thing happens with epuppet and mason. the type detection in vim's scripts.vim takes over completely

**Comment by lelutin** [comment]
ok I found out why calling `:filetype detect` does not do what we hope it to do: when we call it manually, we're not triggering any of the autocmd actions `BufRead` or `BufNewFile` so the autocmd is not processed and we only get vim's built-in type detection.

Now with that info, I need to think of a way to get the tests to trigger those actions so that detection can be tested correctly.

**Comment by lelutin** [comment]
and I found it! with `:edit` without any argument we can trigger autodetection through the expected actions.

oof. I'll just wait for CI to confirm the fix and with that I think the patch series is complete and ready for merge

**Pull request merged**